galasa-parent/dev.galasa.framework.api.openapi.servlet/src/main/java/dev/galasa/framework/api/openapi/servlet/routes/OpenApiRoute.java

```diff
@@ -0,0 +1,110 @@
+import dev.galasa.framework.api.common.QueryParameters;
+import dev.galasa.framework.api.common.ResponseBuilder;
+import dev.galasa.framework.api.common.ServletError;
+import dev.galasa.framework.spi.FrameworkException;
+import dev.galasa.framework.spi.utils.GalasaGson;
+
+/**
+ * This is the base /openapi route that is used to serve the Galasa API server's OpenAPI specification,
+ * either in YAML or JSON format, depending on the MIME type requested.
+ */
+public class OpenApiRoute extends BaseRoute {
+
+    private static final String OPENAPI_SERVER_TEMPLATE = "{API_SERVER_URL}";
+    private static final String OPENAPI_FILE_NAME = "openapi.yaml";
+
+    private static final String APPLICATION_JSON = "application/json";
+    private static final String APPLICATION_YAML = "application/yaml";
+    private static final List<String> SUPPORTED_CONTENT_TYPES = List.of(APPLICATION_YAML, APPLICATION_JSON);
+
+    private static final GalasaGson gson = new GalasaGson();
+
+    private final String openApiYamlContents;
+    private final String openApiJsonContents;
+
+    private Log logger = LogFactory.getLog(this.getClass());
+    private String apiServerUrl;
+
+    public OpenApiRoute(ResponseBuilder responseBuilder, String apiServerUrl) throws IOException {
+        super(responseBuilder, "");
+        this.apiServerUrl = apiServerUrl;
+
+        // The OpenAPI file is small so it can be loaded into memory and served when required
+        // instead of reading it on every request to this endpoint
+        openApiYamlContents = getOpenApiFileAsYamlString();
+        openApiJsonContents = convertYamlToJson(openApiYamlContents);
+    }
+
+    @Override
+    public HttpServletResponse handleGetRequest(String pathInfo, QueryParameters queryParams,
+            HttpServletRequest request, HttpServletResponse response)
+            throws ServletException, IOException, FrameworkException {
+
+        logger.info("OpenApiRoute: handleGetRequest() Entered");
+        String responseContentType = validateContentType(request.getContentType());
+
+        // Convert the YAML contents into a JSON string if the user expects a JSON response
+        String responseContents = openApiYamlContents;
+        if (responseContentType.equals(APPLICATION_JSON)) {
+            responseContents = openApiJsonContents;
+        }
+        return getResponseBuilder().buildResponse(request, response, responseContentType, responseContents, HttpServletResponse.SC_OK);
+    }
+
+    private String getOpenApiFileAsYamlString() throws IOException {
+        logger.info("Reading OpenAPI specification");
+        String contents = "";
+        try (InputStream stream = getClass().getClassLoader().getResourceAsStream(OPENAPI_FILE_NAME)) {
+            contents = new String(stream.readAllBytes());
+
+            // Replace the templated API server URL with the ecosystem's actual API server URL
+            if (apiServerUrl != null) {
```

Comment:
If the apiServerUrl is null, isn't that an error of some sort ? What are the implications ?

Reply:
Added error handling